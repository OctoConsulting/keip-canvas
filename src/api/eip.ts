export interface EipId {
  namespace: string
  name: string
}
