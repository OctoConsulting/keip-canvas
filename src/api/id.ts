export interface EipId {
  namespace: string
  name: string
}

export interface ChildNodeId {
  name: string
  parentEipId: EipId
  parentNodeId: string
}

const childIdToString = (cid: ChildNodeId) =>
  `${cid.parentEipId.namespace}-${cid.parentEipId.name}-${cid.parentNodeId}-${cid.name}`

export const areChildIdsEqual = (c1: ChildNodeId, c2: ChildNodeId) =>
  childIdToString(c1) === childIdToString(c2)
