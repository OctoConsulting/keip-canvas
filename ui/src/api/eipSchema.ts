type RestrictionType = "enum"

interface Restriction {
  type: RestrictionType
  values?: string[]
}

export type AttributeTypes = string | boolean | number

export interface Attribute {
  name: string
  type: "string" | "boolean" | "number"
  description?: string
  default?: string | number | boolean
  required?: boolean
  restriction?: Restriction
}

export type ConnectionType =
  | "passthru"
  | "request_reply"
  | "sink"
  | "source"
  | "tee"

export type Role = "channel" | "endpoint" | "router" | "transformer"

interface Occurrence {
  min?: number
  max?: number
}

interface EipElement {
  name: string
  description?: string
  attributes?: Attribute[]
  childGroup?: EipChildGroup
}

export interface EipChildGroup {
  indicator: "all" | "choice" | "sequence"
  children: EipChildElement[]
  occurrence?: Occurrence
}

export interface EipChildElement extends EipElement {
  occurrence?: Occurrence
}

export interface EipComponent extends EipElement {
  role: Role
  connectionType: ConnectionType
}

export type EipSchema = Record<string, EipComponent[]>
