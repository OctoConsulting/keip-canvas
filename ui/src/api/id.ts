// TODO: use the EipId defined in the API generated from JSON schema instead.
export interface EipId {
  namespace: string
  name: string
}

export interface ChildNodeId {
  name: string
  parentNodeId: string
}

export const childIdToString = (cid: ChildNodeId) =>
  `${cid.parentNodeId}.${cid.name}`

export const areChildIdsEqual = (c1: ChildNodeId, c2: ChildNodeId) =>
  childIdToString(c1) === childIdToString(c2)
