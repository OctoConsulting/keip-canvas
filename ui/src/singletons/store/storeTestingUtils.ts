import { renderHook } from "@testing-library/react"

// Zustand mock import (see '/home/aasaad/projects/keip-canvas/ui/__mocks__/zustand.ts')
// @ts-expect-error the reset function does not exist as a zustand export.
// It is added to the mock as a convenience for setup and cleanup of the store in the tests
import { resetStore } from "zustand"

// WARNING: In test files, import the reset method last, to ensure it's properly initialized.
export const resetMockStore = resetStore as (s: object) => void

export function renderAndUnwrapHook<T>(hookFn: () => T): T {
  const { result } = renderHook(hookFn)
  return result.current
}
