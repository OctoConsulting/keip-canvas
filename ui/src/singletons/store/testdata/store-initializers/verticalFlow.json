{
    "nodes": [
      {
        "id": "SLamPCIOnG",
        "type": "eipNode",
        "position": {
          "x": 2.5,
          "y": 0
        },
        "targetPosition": "top",
        "sourcePosition": "bottom",
        "data": {
          "eipId": {
            "namespace": "integration",
            "name": "inbound-channel-adapter"
          }
        },
        "width": 114,
        "height": 121,
        "positionAbsolute": {
          "x": 2.5,
          "y": 0
        },
        "selected": false,
        "dragging": false
      },
      {
        "id": "UrnBcciyaA",
        "type": "eipNode",
        "position": {
          "x": 2.5,
          "y": 141
        },
        "targetPosition": "top",
        "sourcePosition": "bottom",
        "data": {
          "eipId": {
            "namespace": "integration",
            "name": "header-enricher"
          }
        },
        "width": 114,
        "height": 76,
        "selected": false,
        "positionAbsolute": {
          "x": 2.5,
          "y": 141
        },
        "dragging": false
      },
      {
        "id": "4s1rNUykz1",
        "type": "eipNode",
        "position": {
          "x": 0,
          "y": 237
        },
        "targetPosition": "top",
        "sourcePosition": "bottom",
        "data": {
          "eipId": {
            "namespace": "integration",
            "name": "outbound-channel-adapter"
          }
        },
        "width": 119,
        "height": 121,
        "positionAbsolute": {
          "x": 0,
          "y": 237
        }
      }
    ],
    "edges": [
      {
        "source": "SLamPCIOnG",
        "sourceHandle": "output",
        "target": "UrnBcciyaA",
        "targetHandle": "input",
        "id": "reactflow__edge-SLamPCIOnGoutput-UrnBcciyaAinput"
      },
      {
        "source": "UrnBcciyaA",
        "sourceHandle": "output",
        "target": "4s1rNUykz1",
        "targetHandle": "input",
        "id": "reactflow__edge-UrnBcciyaAoutput-4s1rNUykz1input"
      }
    ],
    "eipNodeConfigs": {
      "SLamPCIOnG": {
        "attributes": {},
        "children": {}
      },
      "UrnBcciyaA": {
        "attributes": {},
        "children": {}
      },
      "4s1rNUykz1": {
        "attributes": {},
        "children": {}
      }
    },
    "selectedChildNode": null,
    "layout": {
      "orientation": "vertical",
      "density": "compact"
    }
  }