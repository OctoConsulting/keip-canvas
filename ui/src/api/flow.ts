import { Edge, Node } from "reactflow"
import { Attribute } from "./generated/eipComponentDef"
import { Attributes } from "./generated/eipFlow"
import { EipId } from "./id"

export interface Layout {
  orientation: "horizontal" | "vertical"
  density: "compact" | "comfortable"
}

export interface EipNodeData {
  eipId: EipId
  label?: string
}

export type EipFlowNode = Node<EipNodeData>

export interface ChannelMapping {
  mapperName: string
  matcher: Attribute
  matcherValue?: string
  isDefaultMapping?: boolean
}

// TODO: Should custom edge data contain only a reference to the mapping?
// The mapping could be stored separately in the app store.
export interface DynamicEdgeData {
  mapping: ChannelMapping
}

export type DynamicEdge = Edge<DynamicEdgeData>

export const EIP_NODE_TYPE = "eipNode"

export const DYNAMIC_EDGE_TYPE = "dynamicEdge"

// TODO: Rethink router interfaces
export interface RouterKeyDef {
  name: string
  type: "attribute" | "child"
  attributesDef: Attribute[]
}

export interface RouterKey {
  name: string
  attributes?: Attributes
}
