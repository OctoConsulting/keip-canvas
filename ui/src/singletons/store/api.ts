import { Edge } from "reactflow"
import { EipFlowNode, Layout, RouterKey } from "../../api/flow"
import { Attributes } from "../../api/generated/eipFlow"
import { EipId } from "../../api/id"

export interface EipConfig {
  attributes: Attributes
  children: string[]
  eipId: EipId
  description?: string
  routerKey?: RouterKey
}

export interface AppStore {
  nodes: EipFlowNode[]
  edges: Edge[]
  eipConfigs: Record<string, EipConfig>
  selectedChildNode: string | null
  layout: Layout
}

export interface SerializedFlow {
  nodes: AppStore["nodes"]
  edges: AppStore["edges"]
  eipConfigs: AppStore["eipConfigs"]
  version: string
}
