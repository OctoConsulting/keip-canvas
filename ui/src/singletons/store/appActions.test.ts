import { act } from "@testing-library/react"
import isDeepEqual from "fast-deep-equal"
import { Position } from "reactflow"
import { beforeEach, describe, expect, test, vi } from "vitest"
import {
  ChannelMapping,
  DynamicEdgeData,
  EIP_NODE_TYPE,
  RouterKey,
} from "../../api/flow"
import { ChildNodeId } from "../../api/id"
import {
  getEdgesView,
  getLayout,
  getNodesView,
  ROOT_PARENT,
  useAppActions,
  useGetChildren,
  useGetContentRouterKey,
  useGetEipAttribute,
  useGetNodeDescription,
  useGetSelectedChildNode,
} from "../store"
import { renderAndUnwrapHook, resetMockStore } from "./storeTestingUtils"
import selectedNodeFlow from "./testdata/singleSelectedNodeFlow.json"
import standardFlow from "./testdata/standardFlow.json"
import verticalFlow from "./testdata/verticalFlow.json"

vi.mock("zustand")

const appActions = renderAndUnwrapHook(useAppActions)

const N1_ID = "9KWCqlIyy7"
const N2_ID = "LoiC2CFbLP"

beforeEach(() => {
  act(() => {
    resetMockStore(standardFlow)
  })
})

const getNode = (id: string) => {
  const nodes = getNodesView()
  const target = nodes.find((n) => n.id === id)
  if (!target) {
    throw new Error(`Could not find node with id: ${id}`)
  }
  return target
}

describe("create dropped node", () => {
  test.each([
    {
      msg: "create with horizontal layout",
      storeState: standardFlow,
      orientation: "horizontal",
      handlePosition: {
        targetPosition: Position.Left,
        sourcePosition: Position.Right,
      },
    },
    {
      msg: "create with vertical layout",
      storeState: verticalFlow,
      orientation: "vertical",
      handlePosition: {
        targetPosition: Position.Top,
        sourcePosition: Position.Bottom,
      },
    },
  ])("$msg", ({ storeState, orientation, handlePosition }) => {
    // Initialize store
    act(() => {
      resetMockStore(storeState)
    })

    // Assert expected layout
    const layout = getLayout()
    expect(layout.orientation).toEqual(orientation)

    const eipId = { namespace: "test", name: "transformer" }
    const position = { x: 5, y: 10 }
    act(() => appActions.createDroppedNode(eipId, position))

    const nodes = getNodesView()
    const node = nodes.find((n) => isDeepEqual(n.data.eipId, eipId))!

    expect(node.position).toEqual(position)
    expect(node.id).toBeTruthy()
    expect(node.type).toEqual(EIP_NODE_TYPE)
    expect(node.targetPosition).toEqual(handlePosition.targetPosition)
    expect(node.sourcePosition).toEqual(handlePosition.sourcePosition)
  })
})

describe("update node label", () => {
  test("update node with unique label success", () => {
    const label = "test-label-123"

    act(() => void appActions.updateNodeLabel(N1_ID, label))

    const target = getNode(N1_ID)
    expect(target.data.label).toEqual(label)
  })

  test("update node with duplicated label -> error", () => {
    let target = getNode(N1_ID)
    const originalLabel = target.data.label

    const duplicatedLabel = "test-router"

    const result = appActions.updateNodeLabel(N1_ID, duplicatedLabel)
    expect(result instanceof Error).toBeTruthy()

    target = getNode(N1_ID)
    expect(target.data.label).toEqual(originalLabel)
  })

  test("update non-existent node is a no-op", () => {
    const before = getNodesView()
    act(() => void appActions.updateNodeLabel("fakeid", "test-label"))
    const after = getNodesView()
    expect(before).toEqual(after)
  })
})

describe("update node description", () => {
  test("update node description success", () => {
    const description = "testing description update"
    act(() => appActions.updateNodeDescription(N1_ID, description))

    const actual = renderAndUnwrapHook(() => useGetNodeDescription(N1_ID))
    expect(actual).toEqual(description)
  })

  test("update non-existent node throws error", () => {
    expect(() =>
      appActions.updateNodeDescription("fakeid", "test")
    ).toThrowError()
  })
})

describe("updating EIP Attributes", () => {
  test.each([
    {
      msg: "add root node attr",
      nodeId: N1_ID,
      parentId: ROOT_PARENT,
      attribute: {
        name: "foo",
        value: "bar",
      },
    },
    {
      msg: "add child node attr",
      nodeId: "poller",
      parentId: N1_ID,
      attribute: {
        name: "child-foo",
        value: "child-bar",
      },
    },
  ])("$msg", ({ nodeId, parentId, attribute }) => {
    act(() =>
      appActions.updateEipAttribute(
        nodeId,
        parentId,
        attribute.name,
        attribute.value
      )
    )

    const actual = renderAndUnwrapHook(() =>
      useGetEipAttribute(nodeId, parentId, attribute.name)
    )
    expect(actual).toEqual(attribute.value)
  })

  test.each([
    {
      msg: "overwrite root node attr",
      nodeId: N2_ID,
      parentId: ROOT_PARENT,
      attribute: {
        name: "phase",
        value: "last",
      },
    },
    {
      msg: "overwrite child node attr",
      nodeId: "poller",
      parentId: N1_ID,
      attribute: {
        name: "fixed-rate",
        value: 500,
      },
    },
  ])("$msg", ({ nodeId, parentId, attribute }) => {
    // assert attribute actually exists before updating
    let actual = renderAndUnwrapHook(() =>
      useGetEipAttribute(nodeId, parentId, attribute.name)
    )
    expect(actual).toBeTruthy()
    expect(actual).not.toEqual(attribute.value)

    act(() =>
      appActions.updateEipAttribute(
        nodeId,
        parentId,
        attribute.name,
        attribute.value
      )
    )

    actual = renderAndUnwrapHook(() =>
      useGetEipAttribute(nodeId, parentId, attribute.name)
    )
    expect(actual).toEqual(attribute.value)
  })

  test.each([
    {
      msg: "update non-existent root node",
      nodeId: "fakeid",
      parentId: ROOT_PARENT,
    },
    {
      msg: "update non-existent parent node",
      nodeId: "poller",
      parentId: "fakeid",
    },
    {
      msg: "update non-existent child node",
      nodeId: "fakeid",
      parentId: N1_ID,
    },
  ])("$msg", ({ nodeId, parentId }) => {
    const attribute = { name: "foo", value: "bar" }

    expect(() =>
      appActions.updateEipAttribute(
        nodeId,
        parentId,
        attribute.name,
        attribute.value
      )
    ).toThrowError()
  })
})

describe("deleting EIP Attributes", () => {
  test.each([
    {
      msg: "delete existing root node attribute success",
      nodeId: N2_ID,
      parentId: ROOT_PARENT,
      attrName: "phase",
    },
    {
      msg: "delete existing child node attribute success",
      nodeId: "poller",
      parentId: N1_ID,
      attrName: "fixed-rate",
    },
  ])("$msg", ({ nodeId, parentId, attrName }) => {
    // assert attribute exists before deleting
    let attribute = renderAndUnwrapHook(() =>
      useGetEipAttribute(nodeId, parentId, attrName)
    )
    expect(attribute).toBeTruthy()

    act(() => appActions.deleteEipAttribute(nodeId, parentId, attrName))

    attribute = renderAndUnwrapHook(() =>
      useGetEipAttribute(nodeId, parentId, attrName)
    )
    expect(attribute).toBeUndefined()
  })

  test.each([
    {
      msg: "delete non-existent root node attribute is a no-op",
      nodeId: N2_ID,
      parentId: ROOT_PARENT,
    },
    {
      msg: "delete non-existent child node attribute is a no-op",
      nodeId: "poller",
      parentId: N1_ID,
    },
  ])("$msg", ({ nodeId, parentId }) => {
    const attrName = "fakeattr"

    act(() => appActions.deleteEipAttribute(nodeId, parentId, attrName))

    const attribute = renderAndUnwrapHook(() =>
      useGetEipAttribute(nodeId, parentId, attrName)
    )
    expect(attribute).toBeUndefined()
  })
})

describe("mapping a dynamic router edge", () => {
  test("update a dynamic routing edge mapping success", () => {
    const edgeId = "reactflow__edge-LoiC2CFbLPoutput-SV43RVeijQinput"
    const mapping: ChannelMapping = {
      mapperName: "test-mapping",
      matcher: { name: "test-match-attr", type: "string" },
      matcherValue: "abc",
    }

    act(() => appActions.updateDynamicEdgeMapping(edgeId, mapping))

    const edge = getEdgesView().find((edge) => edge.id === edgeId)!
    const data = edge.data as DynamicEdgeData
    expect(data.mapping).toEqual(mapping)
  })

  test("update a default edge throws an error", () => {
    const edgeId = "reactflow__edge-9KWCqlIyy7output-LoiC2CFbLPinput"
    const mapping: ChannelMapping = {
      mapperName: "test-mapping",
      matcher: { name: "test-match-attr", type: "string" },
      matcherValue: "abc",
    }

    expect(() =>
      appActions.updateDynamicEdgeMapping(edgeId, mapping)
    ).toThrowError()
  })
})

describe("update content router key", () => {
  test("update existing content router key success", () => {
    const attrName = "expression"
    const attrValue = "bar"
    const routerKey: RouterKey = {
      name: "test-key",
      attributes: { [attrName]: attrValue },
    }

    act(() =>
      appActions.updateContentRouterKey(
        N2_ID,
        routerKey.name,
        attrName,
        attrValue
      )
    )

    const actual = renderAndUnwrapHook(() => useGetContentRouterKey(N2_ID))
    expect(actual).toEqual(routerKey)
  })

  test("add additional router key attribute success", () => {
    const initialKey = renderAndUnwrapHook(() => useGetContentRouterKey(N2_ID))

    const attrName = "foo"
    const attrValue = "bar"
    const routerKey: RouterKey = {
      name: "test-key",
      attributes: { ...initialKey?.attributes, [attrName]: attrValue },
    }

    act(() =>
      appActions.updateContentRouterKey(
        N2_ID,
        routerKey.name,
        attrName,
        attrValue
      )
    )

    const actual = renderAndUnwrapHook(() => useGetContentRouterKey(N2_ID))
    expect(actual).toEqual(routerKey)
  })

  // TODO: unskip this test once action is fixed
  test.skip("update non-router node throws error", () => {
    expect(() =>
      appActions.updateContentRouterKey(N1_ID, "test-key", "foo", "bar")
    ).toThrowError()
  })
})

test("update enabled children success", () => {
  const children = ["child1", "child2"]

  act(() => appActions.updateEnabledChildren(N2_ID, children))

  const actual = renderAndUnwrapHook(() => useGetChildren(N2_ID))
  expect(actual).toEqual(children)
})

test("update selected child node success", () => {
  const childId: ChildNodeId = {
    name: "poller",
    parentNodeId: N2_ID,
  }

  act(() => appActions.updateSelectedChildNode(childId))

  const actual = renderAndUnwrapHook(useGetSelectedChildNode)
  expect(actual).toEqual(childId)
})

test("clear flow success", () => {
  act(() => appActions.clearFlow())

  const nodes = getNodesView()
  const edges = getEdgesView()
  const selectedChildNode = renderAndUnwrapHook(useGetSelectedChildNode)

  expect(nodes).toHaveLength(0)
  expect(edges).toHaveLength(0)
  expect(selectedChildNode).toBeNull()
})

test("clear root node selections from diagram success", () => {
  // Initialize store
  act(() => {
    resetMockStore(selectedNodeFlow)
  })

  // assert a selected node is present before clearing
  let selected = getNodesView().find((n) => n.selected)
  expect(selected).not.toBeUndefined()

  act(() => appActions.clearDiagramSelections())

  selected = getNodesView().find((n) => n.selected)
  expect(selected).toBeUndefined()
})
