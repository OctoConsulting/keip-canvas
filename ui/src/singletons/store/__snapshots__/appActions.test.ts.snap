// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`import flow from an exported JSON file > import a valid flow 1`] = `
{
  "children": [
    {
      "81HAL73ElW": [],
    },
    {
      "Hz22QyhCpj": [],
    },
    {
      "THgI9WvsJh": [
        "transactional",
      ],
    },
  ],
  "edges": [
    {
      "id": "reactflow__edge-81HAL73ElWoutput-Hz22QyhCpjinput",
      "source": "81HAL73ElW",
      "sourceHandle": "output",
      "target": "Hz22QyhCpj",
      "targetHandle": "input",
    },
    {
      "id": "reactflow__edge-Hz22QyhCpjoutput-THgI9WvsJhinput",
      "source": "Hz22QyhCpj",
      "sourceHandle": "output",
      "target": "THgI9WvsJh",
      "targetHandle": "input",
    },
  ],
  "nodes": [
    {
      "data": {
        "eipId": {
          "name": "inbound-gateway",
          "namespace": "web-services",
        },
        "label": "ws-in",
      },
      "dragging": false,
      "height": 124,
      "id": "81HAL73ElW",
      "position": {
        "x": 0,
        "y": 0,
      },
      "positionAbsolute": {
        "x": 0,
        "y": 0,
      },
      "selected": false,
      "sourcePosition": "right",
      "targetPosition": "left",
      "type": "eipNode",
      "width": 128,
    },
    {
      "data": {
        "eipId": {
          "name": "transformer",
          "namespace": "integration",
        },
      },
      "dragging": false,
      "height": 87,
      "id": "Hz22QyhCpj",
      "position": {
        "x": 203,
        "y": 18.5,
      },
      "positionAbsolute": {
        "x": 203,
        "y": 18.5,
      },
      "selected": false,
      "sourcePosition": "right",
      "targetPosition": "left",
      "type": "eipNode",
      "width": 99,
    },
    {
      "data": {
        "eipId": {
          "name": "outbound-gateway",
          "namespace": "http",
        },
      },
      "dragging": false,
      "height": 121,
      "id": "THgI9WvsJh",
      "position": {
        "x": 377,
        "y": 5.5,
      },
      "positionAbsolute": {
        "x": 377,
        "y": 5.5,
      },
      "selected": false,
      "sourcePosition": "right",
      "targetPosition": "left",
      "type": "eipNode",
      "width": 109,
    },
  ],
}
`;
